### PR TITLE
Invalidate its foreign table's relcache when alter server

### PR DESCRIPTION
**Pull request by adam8157**
Foreign table's distribution policy depends on the foreign server's
`mpp_excute` and `num_segments` options, to force update the cached
policy, if the option matches, invalidate the relcache of this server's
every foreign table.

If not, Greenplum still provides the old policy which is stored in the
relation cache since the table's catalog never changed, the generated
plan would be wrong.

### CODE REVIEW

File: src/backend/commands/foreigncmds.c [hunk 1]
```diff
@@ -1144,6 +1159,45 @@ AlterForeignServer(AlterForeignServerStmt *stmt)
...
+
+				scan = table_beginscan_catalog(relation_rel, 0, NULL);
+
+				while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+				{
+					Form_pg_foreign_table ftbl = (Form_pg_foreign_table)GETSTRUCT(tuple);
+
+					if (ftbl->ftserver == srvId)
+						CacheInvalidateRelcacheByRelid(ftbl->ftrelid);
+				}
```
Comment (avamingli) [inline src/backend/commands/foreigncmds.c:1191]: Why not use syscache `SearchSysCacheCopy1(FOREIGNSERVEROID` ?

Comment (adam8157) [reply src/backend/commands/foreigncmds.c:1191]: `SearchSysCache` could only search a specific oid, correct? What I need is to loop the syscache/catlog.

File: src/backend/commands/foreigncmds.c [hunk 2]
```diff
@@ -1144,6 +1159,45 @@ AlterForeignServer(AlterForeignServerStmt *stmt)
...
+			 * invalidate the relcache of this server's every foreign table.
+			 */
+			if (strcmp(od->defname, "mpp_execute") == 0 ||
+				strcmp(od->defname, "num_segments") == 0)
+			{
+				Relation relation_rel;
+				TableScanDesc scan;
+				HeapTuple tuple;
+
+				relation_rel = table_open(ForeignTableRelationId, AccessShareLock);
```
Comment (kainwen) [inline src/backend/commands/foreigncmds.c:1181]: I think we'd better use index scan on `pg_depend` here, just like in psql:

```
gpadmin=# select 1417::regclass::text;
       text
-------------------
 pg_foreign_server
(1 row)

gpadmin=# explain select * from pg_depend where refclassid = 1417 and refobjid = 16979 and refobjsubid = 0;
                                           QUERY PLAN
------------------------------------------------------------------------------------------------
 Index Scan using pg_depend_reference_index on pg_depend  (cost=0.29..8.31 rows=1 width=25)
   Index Cond: ((refclassid = '1417'::oid) AND (refobjid = '16979'::oid) AND (refobjsubid = 0))
 Optimizer: Postgres query optimizer
(3 rows)

gpadmin=# select * from pg_depend where refclassid = 1417 and refobjid = 16979 and refobjsubid = 0;
 classid | objid | objsubid | refclassid | refobjid | refobjsubid | deptype
---------+-------+----------+------------+----------+-------------+---------
    1259 | 16983 |        0 |       1417 |    16979 |           0 | n
(1 row)
```

Comment (adam8157) [reply src/backend/commands/foreigncmds.c:1181]: Why is it better? We clearly know the relationship between foreign servers and tables, scanning the `pg_foreign_table` is straightforward.

`pg_depend` has tons of other things, and a foreign server also has other stuff like user mappings in `pg_depend`, all of these and the index scan make things much more complex I suppose.

Comment (kainwen) [reply src/backend/commands/foreigncmds.c:1181]: > Why is it better? We clearly know the relationship between foreign servers and tables, scanning the `pg_foreign_table` is straightforward.
> 
> `pg_depend` has tons of other things, and a foreign server also has other stuff like user mappings in `pg_depend`, all of these and the index scan make things much more complex I suppose.

Make sense.

`pg_depend` is so huge each point query using index might not be that efficient.

And `alter server` should be very infrequent so seqscan whole pg_foriegin_table is also acceptable.

File: src/backend/commands/foreigncmds.c [hunk 3]
```diff
@@ -1144,6 +1159,45 @@ AlterForeignServer(AlterForeignServerStmt *stmt)
...
+
+				scan = table_beginscan_catalog(relation_rel, 0, NULL);
+
+				while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
```
Comment (kainwen) [inline src/backend/commands/foreigncmds.c:1185]: systable_next?

File: src/backend/commands/foreigncmds.c (hunk 2 above)
Comment (soumyadeep2007) [reply src/backend/commands/foreigncmds.c:1181]: >Make sense.

>pg_depend is so huge each point query using index might not be that efficient.

>And alter server should be very infrequent so seqscan whole pg_foriegin_table is also acceptable.

Agreed.

File: src/backend/commands/foreigncmds.c (hunk 3 above)
Comment (avamingli) [reply src/backend/commands/foreigncmds.c:1185]: Do you mean systable_getnext?

File: src/backend/commands/foreigncmds.c [hunk 4]
```diff
@@ -794,6 +796,19 @@ AlterForeignDataWrapper(AlterFdwStmt *stmt)
...
+		{
+			DefElem *od = lfirst(optcell);
+
+			if (strcmp(od->defname, "mpp_execute") == 0)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("\"%s\" of foreign data wrapper is not allowed to be altered", od->defname)));
+			}
+		}
```
Comment (soumyadeep2007) [inline src/backend/commands/foreigncmds.c:810]: We don't need `if (stmt->options)` as foreach handles empty lists. So we can remove the outer if.

File: src/backend/commands/foreigncmds.c [hunk 5]
```diff
@@ -794,6 +796,19 @@ AlterForeignDataWrapper(AlterFdwStmt *stmt)
...
+		foreach(optcell, stmt->options)
+		{
+			DefElem *od = lfirst(optcell);
+
+			if (strcmp(od->defname, "mpp_execute") == 0)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("\"%s\" of foreign data wrapper is not allowed to be altered", od->defname)));
+			}
```
Comment (soumyadeep2007) [inline src/backend/commands/foreigncmds.c:809]: This logic belongs in a separate commit. Please don't add it to the relcache fix commit.

**Review by soumyadeep2007 — COMMENTED** [review]
Two things:

(1) Please follow the pattern used in `AlterPublication` with respect to resetting full relcache in case we exceed MAX_RELCACHE_INVAL_MSGS (can happen if we have many foreign partitions for instance).

So something like:
```
List	   *relids = GetForeignServerRelations(<server_oid>); // similar to GetPublicationRelations

/*
 * We don't want to send too many individual messages, at some point
 * it's cheaper to just reset whole relcache.
 */
if (list_length(relids) < MAX_RELCACHE_INVAL_MSGS)
{
	ListCell   *lc;
	foreach(lc, relids)
	{
		Oid			relid = lfirst_oid(lc);
		CacheInvalidateRelcacheByRelid(relid);
	}
}
else
	CacheInvalidateRelcacheAll();
```

(2) Please add a breaking isolation2 test and include the ERROR message we were seeing in the commit message. @kainwen was already showing me offline a repro with an open transaction and a concurrent invocation of `ALTER FOREIGN SERVER`.

File: src/backend/commands/foreigncmds.c (hunk 3 above)
Comment (adam8157) [reply src/backend/commands/foreigncmds.c:1185]: The common practice in the code base seems like this:

`table_beginscan_catalog()` and `heap_getnext()` are for normal catalog scan, `systable_beginscan()` and `systable_getnext()` are for catalog scan with index

File: src/backend/commands/foreigncmds.c (hunk 5 above)
Comment (jingwen-yang-yjw) [reply src/backend/commands/foreigncmds.c:809]: And I notice that it's unable to specify option `mpp_execute` when CREATE/ALTER EXTENSION.
(No such syntax for CREATE/ALTER EXTENSION.)
So this logic is not needed actually.

```
testdb=# create extension postgres_fdw options (mpp_execute 'all segments');
ERROR:  syntax error at or near "options"
LINE 1: create extension postgres_fdw options (mpp_execute 'all segm...
                                      ^
```

File: src/backend/commands/foreigncmds.c (hunk 4 above)
Comment (adam8157) [reply src/backend/commands/foreigncmds.c:810]: Oh, it was there for the existing alter options workflow.

File: src/backend/commands/foreigncmds.c (hunk 5 above)
Comment (adam8157) [reply src/backend/commands/foreigncmds.c:809]: They are related, but for sure a separate commit and message explaining it would be better, thanks.

Comment (adam8157) [reply src/backend/commands/foreigncmds.c:809]: @jingwen-yang-yjw `CREATE EXTENSION` actually calls a SQL file, the FOREIGN WRAPPER commands are in it.

**Review by kainwen — APPROVED** [review]
Code change LGTM. I don't think we need an islolatione2 case because it needs postgres fdw, and we already have cases. Not sure if possible to add one in isolation2.

Others please resolve @soumyadeep2007 's comments.

Thanks all!
